### PR TITLE
Fix rare deadlock on multiple concurrent conn.Close() calls.

### LINKED ISSUE #226: Concurrent Close calls to Connection struct may result in deadlock.
In rare situations, only first goroutine will successfully get the result inside `func (me *Connection) call`. Others will wait indefinitely on `select`.

### PR DESCRIPTION
**Pull request by DmitriyMV**
Fix #226

### CODE REVIEW

**Comment by michaelklishin** [comment]
So this seems to fix the same issue as https://github.com/streadway/amqp/pull/196. We'll need some time to weigh the two and merge the preferred version. Thank you.

**Comment by DmitriyMV** [comment]
If I understand #196 correctly - it's about concurrent access from Channels. This fix is about sharing Connection object. Still - I agree with you, and would rather wait for other merge request to pass checks and see how this will affect the problem.

Will investigate again when #196 is green.

**Comment by michaelklishin** [comment]
@DmitriyMV OK, we will take a look at this PR earlier then. I've seen Travis failures that I cannot reproduce locally so it's not clear if the changes in #196 are to blame :(

File: connection_test.go [hunk 1]
```diff
@@ -34,3 +35,23 @@ func TestQueueDeclareOnAClosedConnectionFails(t *testing.T) {
...
 }
+
+func TestConcurrentClose(t *testing.T) {
+	conn, err := Dial("amqp://guest:guest@localhost:5672/")
+	if err != nil {
+		log.Fatalf("Could't connect to amqp server, err = %s", err)
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(5)
```
Comment (michaelklishin) [inline connection_test.go]: When I change this value to something higher than 5, I cannot get the suite to pass 20 times in a row, e.g.

```
for i in {1..20}; do go test -v -tags integration_test.go .\/...; done
```

Comment (michaelklishin) [inline connection_test.go]: Never mind, I figured it out.

**Comment by michaelklishin** [comment]
Thank you!

**Pull request merged**